manager.go

```diff
@@ -1569,23 +1585,87 @@ func (m *Manager) sanityCheck() error {
-// instantiate the connection at the manager level and cache the netlink socket.
-func (m *Manager) newNetlinkConnection(ifindex int32, netns uint64) (*netlinkCacheValue, error) {
-	var cacheEntry netlinkCacheValue
-	var err error
-	// Open a netlink socket for the requested namespace
-	cacheEntry.rtNetlink, err = tc.Open(&tc.Config{
-		NetNS: int(netns),
-	})
+// NewCachedNetlinkSocket - TC classifiers are attached by creating a qdisc on the requested interface. A netlink socket
+// is required to create a qdisc (or to attach an XDP program to an interface). Since this socket can be re-used for
+// multiple probes, instantiate the connection at the manager level and cache the netlink socket.
+func (m *Manager) NewCachedNetlinkSocket(nsHandle uint64, nsID uint32) (*NetlinkSocket, error) {
+	m.stateLock.Lock()
+	defer m.stateLock.Unlock()
+	if m.state < initialized {
+		return nil, ErrManagerNotInitialized
+	}
+	return m.newCachedNetlinkSocket(nsHandle, nsID)
+}
+
+// newCachedNetlinkSocket - Internal function (see NewCachedNetlinkSocket)
+func (m *Manager) newCachedNetlinkSocket(nsHandle uint64, nsID uint32) (*NetlinkSocket, error) {
+	cacheEntry, err := NewNetlinkSocket(nsHandle, nsID)
 	if err != nil {
-		return nil, fmt.Errorf("couldn't open a NETLink socket in namespace %v: %w", netns, err)
+		return nil, err
 	}
 
 	// Insert in manager cache
-	m.netlinkCache[netlinkCacheKey{Ifindex: ifindex, Netns: netns}] = &cacheEntry
-	return &cacheEntry, nil
+	m.netlinkSocketCache[nsID] = cacheEntry
+	return cacheEntry, nil
+}
+
+// GetNetlinkSocket - Returns a netlink socket in the requested network namespace from cache or creates a new one.
+func (m *Manager) GetNetlinkSocket(nsHandle uint64, nsID uint32) (*NetlinkSocket, error) {
+	m.stateLock.Lock()
+	defer m.stateLock.Unlock()
+	if m.state < initialized {
+		return nil, ErrManagerNotInitialized
+	}
+	return m.getNetlinkSocket(nsHandle, nsID)
+}
+
+// getNetlinkSocket - Internal function (see GetNetlinkSocket)
+func (m *Manager) getNetlinkSocket(nsHandle uint64, nsID uint32) (*NetlinkSocket, error) {
+	sock, ok := m.netlinkSocketCache[nsID]
+	if ok {
+		return sock, nil
+	}
+
+	return m.newCachedNetlinkSocket(nsHandle, nsID)
+}
+
+// CleanupNetworkNamespace - Cleans up all references to the provided network namespace within the manager. This means
+// that any TC classifier or XDP probe in that network namespace will be stopped and all opened netlink socket in that
+// namespace will be closed.
+// WARNING: Don't forget to call this method if you've provided a IfIndexNetns and IfIndexNetnsID to one of the probes
+// of this manager. Failing to call this cleanup function may lead to leaking the network namespace. Only call this
+// function when you're sure that the manager no longer needs to perform anything in the provided network namespace (or
+// else call NewNetlinkSocket first).
+func (m *Manager) CleanupNetworkNamespace(nsID uint32) error {
+	var err error
```

Comment:
Does this need to grab the `stateLock` mutex?

Reply:
Good catch, and yes definitely. ✅